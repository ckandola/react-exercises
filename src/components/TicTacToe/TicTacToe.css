.tictactoe-board {
 display: table;
 margin: auto;
}

.tictactoe-board-row:after {
    clear:both;
    content:"";
    display:table;
}

.tictactoe-gameover-button {
    display:block;
    margin:0.5rem auto;
    width: fit-content;
    border-radius: 0.75rem;
    box-shadow: var(--light-shadow);
    padding: 0.1rem 0.75rem;
}
