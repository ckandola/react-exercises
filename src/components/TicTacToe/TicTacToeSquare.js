import React, {useState, useEffect} from 'react';
import './TicTacToeSquare.css';

const TicTacToeSquare = ({disabled, player, onClick}) => {
  const [value, setValue] = useState("");

  const handleClick = () => {
    setValue(player);
    onClick();
  }

  useEffect(() => {
    if (!disabled) {
      setValue("");
    }
  }, [disabled]);

    return (
      <button className="tictactoe-square" disabled={disabled} onClick={handleClick}>{value}</button>
    );
}


export default TicTacToeSquare;