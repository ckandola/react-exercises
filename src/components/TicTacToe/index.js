import React , { useState } from 'react';
import TicTacToeSquare from './TicTacToeSquare';
import './TicTacToe.css';

const TicTacToe = () => {
  const [count, setCount] = useState(0);
  const [usedSpots, setUsedSpots] = useState(new Array(9));
  const [gameOver, setGameOver] = useState(false);
  const [gameOverMessage, setGameOverMessage] = useState('GAME OVER');

  const playAgain = () => {
    setGameOverMessage('GAME OVER');
    setGameOver(false);
    setCount(0);
    setUsedSpots([]);
  }

  const calculateWin = spots => {
    const wins = [
      [0, 3, 6],
      [1, 4, 7],
      [2, 5, 8],
      [0, 1, 2],
      [3, 4, 5],
      [6, 7, 8],
      [0, 4, 8],
      [2, 4, 6]   
    ];
    for(let i = 0; i < wins.length; i++) {
      const [a, b, c] = wins[i];
      if (spots[a] && spots[a] === spots[b] && spots[a] === spots[c]) {
        setGameOver(true);
        setGameOverMessage(player === 'X' ? gameOverMessage + ' - X WINS' : gameOverMessage + ' - O WINS');
        break;
      }
    }
  }

  const incrementCount = id => {
    if(count === 8) {
      setGameOverMessage(gameOverMessage + ' - NO WINNER');
      setGameOver(true);
    }
    setCount(c => c + 1);
      const spotsCopy = usedSpots.slice();
      spotsCopy[id] = player;
      setUsedSpots(spotsCopy);
      calculateWin(spotsCopy);
  }

  const player = count % 2 === 0 ? 'X' : 'O';
  return (
    <div className="board">
      <div className="message" key="turn">Turn: {count}</div>
      {!gameOver ? 
        <div key="player">Player: {player}</div> :
        <div> 
          {gameOverMessage}
          <div>
            <button className="tictactoe-gameover-button" onClick={playAgain}>Play Again</button>
          </div>
        </div>
      }
      <div className="board-row" key={`row-1`}>
        <TicTacToeSquare
          key={0}
          player={player}
          disabled={gameOver || usedSpots[0]}
          onClick={() => incrementCount(0)}
        />
        <TicTacToeSquare
          key={1}
          player={player}
          disabled={gameOver || usedSpots[1]}
          onClick={() => incrementCount(1)}
        />
        <TicTacToeSquare
          key={2}
          player={player}
          disabled={gameOver || usedSpots[2]}
          onClick={() => incrementCount(2)}
        />
      </div>
      <div className="board-row" key={`row-2`}>
        <TicTacToeSquare
          key={3}
          player={player}
          disabled={gameOver || usedSpots[3]}
          onClick={() => incrementCount(3)}
        />
        <TicTacToeSquare
          key={4}
          player={player}
          disabled={gameOver || usedSpots[4]}
          onClick={() => incrementCount(4)}
        />
        <TicTacToeSquare
          key={5}
          player={player}
          disabled={gameOver || usedSpots[5]}
          onClick={() => incrementCount(5)}
        />
      </div>
      <div className="board-row" key={`row-3`}>
        <TicTacToeSquare
          key={6}
          player={player}
          disabled={gameOver || usedSpots[6]}
          onClick={() => incrementCount(6)}
        />
        <TicTacToeSquare
          key={7}
          player={player}
          disabled={gameOver || usedSpots[7]}
          onClick={() => incrementCount(7)}
        />
        <TicTacToeSquare
          key={8}
          player={player}
          disabled={gameOver || usedSpots[8]}
          onClick={() => incrementCount(8)}
        />
      </div>
    </div>
  );
}

export default TicTacToe;