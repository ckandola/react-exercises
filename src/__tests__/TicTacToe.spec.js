import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import TicTacToe from '../components/TicTacToe';

describe("TicTacToe tests", () => {
    it('matches the snapshot', () => {
        const wrapper = shallow(<TicTacToe />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('increments turn, changes player to O, and marks the middle square with an "X" on click', () => {
        const wrapper = shallow(<TicTacToe />);
        wrapper.find("TicTacToeSquare").at(4).simulate('click');
        expect(wrapper.find('.tictactoe-message').getElement().props.children).toEqual(["Turn: ", 1]);
        expect(wrapper.find('.tictactoe-player').getElement().props.children).toEqual(["Player: ", "O"]);
        expect(wrapper.find('TicTacToeSquare').at(4).getElement().props.disabled).toBe('X');
    });

    it('declares X the winner when order is 0, 3, 1, 4, 2', () => {
        const wrapper = shallow(<TicTacToe />);
        wrapper.find("TicTacToeSquare").at(0).simulate('click');
        wrapper.find("TicTacToeSquare").at(3).simulate('click');
        wrapper.find("TicTacToeSquare").at(1).simulate('click');
        wrapper.find("TicTacToeSquare").at(4).simulate('click');
        wrapper.find("TicTacToeSquare").at(2).simulate('click');
        expect(wrapper.find(".tictactoe-gameover-msg")).toHaveLength(1);
        expect(wrapper.find(".tictactoe-gameover-msg").getElement().props.children).toBe("GAME OVER - X WINS");
    });

    it('declares no winner if nobody has won by the 9th turn', () => {
        const wrapper = shallow(<TicTacToe />);
        wrapper.find("TicTacToeSquare").at(0).simulate('click');
        wrapper.find("TicTacToeSquare").at(1).simulate('click');
        wrapper.find("TicTacToeSquare").at(3).simulate('click');
        wrapper.find("TicTacToeSquare").at(4).simulate('click');
        wrapper.find("TicTacToeSquare").at(2).simulate('click');
        wrapper.find("TicTacToeSquare").at(6).simulate('click');
        wrapper.find("TicTacToeSquare").at(5).simulate('click');
        wrapper.find("TicTacToeSquare").at(8).simulate('click');
        wrapper.find("TicTacToeSquare").at(7).simulate('click');
        expect(wrapper.find(".tictactoe-gameover-msg").getElement().props.children).toBe("GAME OVER - NO WINNER");
    });

    it('clears the board when Play Again is clicked (after a game is over)', () => {
        const wrapper = shallow(<TicTacToe />);
        wrapper.find("TicTacToeSquare").at(0).simulate('click');
        wrapper.find("TicTacToeSquare").at(3).simulate('click');
        wrapper.find("TicTacToeSquare").at(1).simulate('click');
        wrapper.find("TicTacToeSquare").at(4).simulate('click');
        wrapper.find("TicTacToeSquare").at(2).simulate('click');
        wrapper.find(".tictactoe-gameover-button").simulate('click');
        expect(wrapper.find(".tictactoe-message").getElement().props.children).toEqual(["Turn: ", 0]);
        expect(wrapper.find('[disabled]')).toHaveLength(0);
    });
});
