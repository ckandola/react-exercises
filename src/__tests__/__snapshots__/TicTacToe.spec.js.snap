// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicTacToe tests matches the snapshot 1`] = `
<div
  className="tictactoe-board"
>
  <div
    className="tictactoe-message"
    key="turn"
  >
    Turn: 
    0
  </div>
  <div
    className="tictactoe-player"
  >
    Player: 
    X
  </div>
  <div
    className="tictactoe-board-row"
    key="row-1"
  >
    <TicTacToeSquare
      key="0"
      onClick={[Function]}
      player="X"
    />
    <TicTacToeSquare
      key="1"
      onClick={[Function]}
      player="X"
    />
    <TicTacToeSquare
      key="2"
      onClick={[Function]}
      player="X"
    />
  </div>
  <div
    className="tictactoe-board-row"
    key="row-2"
  >
    <TicTacToeSquare
      key="3"
      onClick={[Function]}
      player="X"
    />
    <TicTacToeSquare
      key="4"
      onClick={[Function]}
      player="X"
    />
    <TicTacToeSquare
      key="5"
      onClick={[Function]}
      player="X"
    />
  </div>
  <div
    className="tictactoe-board-row"
    key="row-3"
  >
    <TicTacToeSquare
      key="6"
      onClick={[Function]}
      player="X"
    />
    <TicTacToeSquare
      key="7"
      onClick={[Function]}
      player="X"
    />
    <TicTacToeSquare
      key="8"
      onClick={[Function]}
      player="X"
    />
  </div>
</div>
`;
